File: xml-sdk/src/test/java/com/cleanster/xml/UsersTest.java
package com.cleanster.xml;

import com.cleanster.xml.api.UsersXmlApi;
import com.cleanster.xml.client.XmlConverter;
import com.cleanster.xml.client.XmlHttpClient;
import com.cleanster.xml.model.User;
import com.cleanster.xml.model.XmlApiResponse;
import org.junit.jupiter.api.*;

import java.util.Map;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.ArgumentMatchers.*;
import static org.mockito.Mockito.*;

class UsersTest {

    private XmlHttpClient http;
    private UsersXmlApi   api;

    private static final String USER_JSON = "{\"success\":true,\"message\":\"OK\","
            + "\"data\":{\"id\":42,\"email\":\"alice@example.com\",\"firstName\":\"Alice\"}}";
    private static final String TOKEN_JSON = "{\"success\":true,\"message\":\"OK\","
            + "\"data\":{\"id\":42,\"token\":\"jwt-abc-123\"}}";
    private static final String VERIFY_JSON = "{\"success\":true,\"message\":\"valid\",\"data\":{}}";

    @BeforeEach void setUp() {
        XmlHttpClient real = XmlHttpClient.builder().baseUrl("http://dummy").accessKey("key").build();
        http = spy(real);
        api  = new UsersXmlApi(http);
    }

    // ── createUser ─────────────────────────────────────────────────────────────

    @Test void createUser_callsPost_v1UserAccount() {
        doReturn(USER_JSON).when(http).post(eq("/v1/user/account"), any());
        api.createUser("alice@example.com", "Alice", "Smith");
        verify(http).post(eq("/v1/user/account"), any());
    }

    @Test void createUser_returnsParsedUser() {
        doReturn(USER_JSON).when(http).post(eq("/v1/user/account"), any());
        XmlApiResponse<User> resp = api.createUser("alice@example.com", "Alice", "Smith");
        assertTrue(resp.isSuccess());
        assertEquals(42, (int) resp.getData().getId());
    }

    @Test void createUser_mapOverload_callsPost() {
        doReturn(USER_JSON).when(http).post(eq("/v1/user/account"), any());
        Map<String, Object> body = Map.of("email", "alice@example.com");
        api.createUser(body);
        verify(http).post(eq("/v1/user/account"), any());
    }

    // ── fetchAccessToken ───────────────────────────────────────────────────────

    @Test void fetchAccessToken_callsGet_v1UserToken() {
        doReturn(TOKEN_JSON).when(http).get("/v1/user/access-token/42");
        api.fetchAccessToken(42);
        verify(http).get("/v1/user/access-token/42");
    }

    @Test void fetchAccessToken_returnsParsedToken() {
        doReturn(TOKEN_JSON).when(http).get("/v1/user/access-token/42");
        XmlApiResponse<User> resp = api.fetchAccessToken(42);
        assertTrue(resp.isSuccess());
        assertEquals("jwt-abc-123", resp.getData().getToken());
    }

    @Test void fetchAccessToken_differentId_callsCorrectPath() {
        doReturn(TOKEN_JSON).when(http).get("/v1/user/access-token/99");
        api.fetchAccessToken(99);
        verify(http).get("/v1/user/access-token/99");
    }

    // ── verifyJwt ──────────────────────────────────────────────────────────────

    @Test void verifyJwt_callsPost_v1UserVerifyJwt() {
        doReturn(VERIFY_JSON).when(http).post(eq("/v1/user/verify-jwt"), any());
        api.verifyJwt("my-jwt-token");
        verify(http).post(eq("/v1/user/verify-jwt"), any());
    }

    @Test void verifyJwt_returnsSuccess() {
        doReturn(VERIFY_JSON).when(http).post(eq("/v1/user/verify-jwt"), any());
        assertTrue(api.verifyJwt("my-jwt-token").isSuccess());
    }

    // ── JAXB ───────────────────────────────────────────────────────────────────

    @Test void user_toXml_isValidXml() {
        User u = new User(); u.setId(1); u.setEmail("test@example.com");
        assertTrue(XmlConverter.isXml(XmlConverter.toXml(u)));
    }

    @Test void user_fromXml_roundTrip() {
        User o = new User(); o.setId(7); o.setEmail("bob@example.com"); o.setToken("tok-xyz");
        User r = XmlConverter.fromXml(XmlConverter.toXml(o), User.class);
        assertEquals(7, (int) r.getId());
        assertEquals("bob@example.com", r.getEmail());
        assertEquals("tok-xyz", r.getToken());
    }
}

File: kotlin-sdk/src/test/kotlin/com/cleanster/UsersTest.kt
package com.cleanster

import kotlinx.coroutines.test.runTest
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class UsersTest {

    private lateinit var mock: MockHttpEngine
    private lateinit var client: CleansterClient

    @BeforeEach fun setUp() {
        mock   = MockHttpEngine()
        client = testClient(mock)
    }

    @Test fun `createUser sends POST`() = runTest {
        mock.succeed(mapOf("id" to 1, "email" to "a@b.com"))
        client.users.createUser("a@b.com", "A", "B")
        assertEquals("POST", mock.capturedMethod)
    }

    @Test fun `createUser sends correct path`() = runTest {
        mock.succeed(mapOf("id" to 1))
        client.users.createUser("a@b.com", "A", "B")
        assertTrue(mock.capturedUrl?.endsWith("/v1/user/account") == true)
    }

    @Test fun `createUser sends access-key header`() = runTest {
        mock.succeed(mapOf("id" to 1))
        client.users.createUser("a@b.com", "A", "B")
        assertEquals("test-key", mock.capturedHeaders?.get("access-key"))
    }

    @Test fun `createUser encodes email`() = runTest {
        mock.succeed(mapOf("id" to 1))
        client.users.createUser("alice@example.com", "Alice", "Smith")
        assertEquals("alice@example.com", mock.capturedBody?.get("email"))
    }

    @Test fun `createUser encodes firstName`() = runTest {
        mock.succeed(mapOf("id" to 1))
        client.users.createUser("a@b.com", "Alice", "Smith")
        assertEquals("Alice", mock.capturedBody?.get("firstName"))
    }

    @Test fun `createUser encodes lastName`() = runTest {
        mock.succeed(mapOf("id" to 1))
        client.users.createUser("a@b.com", "Alice", "Smith")
        assertEquals("Smith", mock.capturedBody?.get("lastName"))
    }

    @Test fun `createUser encodes optional phone`() = runTest {
        mock.succeed(mapOf("id" to 1))
        client.users.createUser("a@b.com", "A", "B", phone = "+14155551234")
        assertEquals("+14155551234", mock.capturedBody?.get("phone"))
    }

    @Test fun `createUser decodes response id`() = runTest {
        mock.succeed(mapOf("id" to 42.0, "email" to "a@b.com"))
        val resp = client.users.createUser("a@b.com", "A", "B")
        assertEquals(42, resp.data?.id)
    }

    @Test fun `fetchAccessToken sends GET`() = runTest {
        mock.succeed(mapOf("token" to "jwt-abc"))
        client.users.fetchAccessToken(123)
        assertEquals("GET", mock.capturedMethod)
    }

    @Test fun `fetchAccessToken sends correct path`() = runTest {
        mock.succeed(mapOf("token" to "jwt-abc"))
        client.users.fetchAccessToken(123)
        assertTrue(mock.capturedUrl?.endsWith("/v1/user/access-token/123") == true)
    }

    @Test fun `fetchAccessToken decodes token`() = runTest {
        mock.succeed(mapOf("token" to "eyJhbGciOi"))
        val resp = client.users.fetchAccessToken(123)
        assertEquals("eyJhbGciOi", resp.data?.token)
    }

    @Test fun `fetchAccessToken interpolates userId`() = runTest {
        mock.succeed(mapOf("token" to "jwt"))
        client.users.fetchAccessToken(9999)
        assertTrue(mock.capturedUrl?.contains("9999") == true)
    }

    @Test fun `verifyJwt sends POST`() = runTest {
        mock.succeedEmpty()
        client.users.verifyJwt("jwt-token")
        assertEquals("POST", mock.capturedMethod)
    }

    @Test fun `verifyJwt sends correct path`() = runTest {
        mock.succeedEmpty()
        client.users.verifyJwt("jwt-token")
        assertTrue(mock.capturedUrl?.endsWith("/v1/user/verify-jwt") == true)
    }

    @Test fun `verifyJwt encodes token`() = runTest {
        mock.succeedEmpty()
        client.users.verifyJwt("my-secret-jwt")
        assertEquals("my-secret-jwt", mock.capturedBody?.get("token"))
    }

    @Test fun `setToken updates client token`() = runTest {
        client.setToken("new-jwt")
        assertEquals("new-jwt", client.getToken())
    }

    @Test fun `token is sent as header`() = runTest {
        client.setToken("user-jwt-123")
        mock.succeed(mapOf("id" to 1))
        client.users.createUser("a@b.com", "A", "B")
        assertEquals("user-jwt-123", mock.capturedHeaders?.get("token"))
    }
}
